[TypeChecker] Don't convert specialzed call into EnumElement pattern

rintaro opened the pull request:
Fixes [SR-3452](https://bugs.swift.org/browse/SR-3452)

Previously, on `case EnumTy.CaseVal<SomeType>():`, `<SomeType>` was silently discarded.
We don't have to convert it into `EnumElementPattern` in the first place.

**Note:**
This was obviously a bug, but the change is technically a source breaking change.
`isSwiftVersion3()` guard is required?

slavapestov commented:
@rintaro Do you have a testcase that used to typecheck and now doesn't? Did the formerly-passing testcase also make it all the way through SILGen and IRGen?

rintaro commented:
@slavapestov 
Added test case used to compile.

`case E.A<T>():` was parsed as:
```
        (case_stmt
          (case_label_item
          (pattern_expr
            (call_expr type='<null>'  arg_labels=
              (unresolved_specialize_expr type='<null>'
                (unresolved_dot_expr type='<null>' field 'A' function_ref=unapplied
                  (unresolved_decl_ref_expr type='<null>' name=E specialized=no) function_ref=unapplied)
                (type_ident
                  (component id='UndefinedType' bind=none)))
              (tuple_expr type='()' location=test.swift:7:28 range=[test.swift:7:28 - line:7:29]))))
```
and then, typechecked as:
```
        (case_stmt
          (case_label_item
          (pattern_enum_element type='E' E.A
            (pattern_tuple type='()' names=)))
```
Since the whole expression used to be wrongfully replaced with `EnumElementPattern`,
SILGen and IRGen didn't catch this.

slavapestov commented:
In that case please put the fix behind isSwiftVersion3(), and add two tests -- test/Parse/enum_element_pattern_swift4.swift that passes -swift-version 4 and asserts that an error was diagnosed, and test/Compatibility/enum_element_pattern.swift that verifies that the code typechecks. In each test, add a comment at the top of the file mentioning the other test.

rintaro commented:
@slavapestov 👍  done.

@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented on test/Parse/enum_element_pattern_swift4.swift:
```diff
@@ -0,0 +1,21 @@
+    break
+  case E.B<Int>(): // expected-error {{cannot specialize a non-generic definition}} expected-note {{while parsing this '<' as a type parameter bracket}}
+    break
+  case E.C(): // Ok.
```
BTW, I'm not sure whether this is correct behavior or not.
`let e: E = E.C()` doesn't typecheck: `error: member 'C' is not a function`

slavapestov replied:
Yeah, I don't think the redundant () should be allowed in Swift 4 mode either.

I notice that redundant () with swift works, but with guard case, a SILGen assert is triggered. I have an old radar on me that says that this typechecks but shouldn't -- try compiling it with -emit-silgen, it will crash:

```
enum E {
  case A, B, C
}

func testE(e: E) {
  switch e {
  case E.A<UndefinedTy>(): // expected-error {{use of undeclared type 'UndefinedTy'}}
    break
  case E.B<Int>(): // expected-error {{cannot specialize a non-generic definition}} expected-note {{while parsing this '<' as a type parameter bracket}}
    break
  case E.C(): // Ok.
    break
  default:
    break
  }
}
```

slavapestov replied:
If you want to really beat this bug to death, you could have Sema drop the redundant () from the AST in this case, so that the `guard case` example doesn't assert.

In a no-assert build, like Swift 3.0 GM, the code even IRGen's OK, so there must be someone out there accidentally relying on this behavior...

rintaro replied:
Thanks!
Unfortunately, this compiles in Swift3.  
```swift
enum E {
  case A, B, C
}

func testE(e: E) {
  switch e {
  case E.C(let payload):
    print(payload) // prints "()"
  default:
    break
  }
}
```
For this case, I think we can't drop `(let payload)` from AST in Swift3 mode.

```swift
enum E {
  case A, B, C
}

func testE(e: E) {
  guard case E.C(let payload) = e { return }
  print(payload)
}
```
does assert-fail in SILGen.
In no-assert Swift3 GM, IRGen (Verifer) emits:
```
Global is external, but doesn't have external or weak linkage!
i8* ()* @_TFF4test5testEFT1eOS_1E_T_auL_7payloadT_
invalid linkage type for function declaration
i8* ()* @_TFF4test5testEFT1eOS_1E_T_auL_7payloadT_
LLVM ERROR: Broken module found, compilation aborted!
```

slavapestov replied:
Well, maybe then we should diagnose these constructs in swift 3 mode too, since emitting broken llvm IR is no good. We are ok with breaking compatibility in cases like this where the old behavior was clearly wrong.

slavapestov replied:
@jrose-apple Thoughts?

jrose-apple replied:
- I think the level of "clearly wrong" these days has to be "actively harmful", whereas this is "silly but the code still worked".
- We are free to change the behavior of source code that previously crashed with asserts off. :-)

If it's not hard, I suggest emitting a warning in Swift 3 mode and an error in Swift 4 mode.

jrose-apple replied:
(I didn't actually look at your implementation.)

rintaro commented:
For [SR-3452](https://bugs.swift.org/browse/SR-3452), Added warning with fix-it in Swift3 mode.

As for `E.C()` cases, I've filed [SR-3466](https://bugs.swift.org/browse/SR-3466) and added `FIXME` test cases.
It seems it's a little too difficult for me...

rintaro commented:
@swift-ci Please smoke test
